paasta_tools/mesos_maintenance.py

```diff
@@ -334,6 +379,118 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    :param mesos_secrets: optional argument specifying the path to the file containing the mesos-slave credentials
+    :returns: a string containing the secret/password
+    """
+    return load_credentials(mesos_secrets)[1]
+
+
+def reserve(slave_id, resource, amount):
+    """Dynamically reserve resources in marathon to prevent tasks from using them.
+    :param slave_id: the id of the mesos slave
+    :param resource: the mesos resource type we want to reserve (i.e. cpus, mem, gpus, etc)
+    :param amount: integer representing the amount of the above resource to reserve
+    :returns: boolean where 0 represents success and 1 is a failure
+    """
+    log.info("Dynamically reserving %f %s on %s" % (amount, resource, slave_id))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(slave_id, resource, amount)).replace("'", '"').replace('+', '%20')
+    }
+    client_fn = reserve_api()
+    try:
+        reserve_output = client_fn(method="POST", endpoint="", data=payload).text
+    except HTTPError as e:
+        e.msg = "Error adding dynamic reservation. Got error: %s" % e.msg
+        raise
+    return reserve_output
+
+
+def unreserve(slave_id, resource, amount):
+    """Dynamically unreserve resources in marathon to allow tasks to using them.
+    :param slave_id: the id of the mesos slave
+    :param resource: the mesos resource type we want to unreserve (i.e. cpus, mem, gpus, etc)
+    :param amount: integer representing the amount of the above resource to unreserve
+    :returns: boolean where 0 represents success and 1 is a failure
+    """
+    log.info("Dynamically unreserving %d %s on %s" % (amount, resource, slave_id))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(slave_id, resource, amount)).replace("'", '"').replace('+', '%20')
+    }
+    client_fn = unreserve_api()
+    try:
+        unreserve_output = client_fn(method="POST", endpoint="", data=payload).text
+    except HTTPError as e:
+        e.msg = "Error adding dynamic unreservation. Got error: %s" % e.msg
+        raise
+    return unreserve_output
+
+
+def reserve_all_resources(hostnames):
+    mesos_state = get_mesos_state_summary_from_leader()
+    hosts = []
+    for hostname in hostnames:
+        if '|' in hostname:
+            (host, _) = hostname.split('|')
+            hosts.append(host)
+        else:
+            hosts.append(hostname)
+    for slave in mesos_state['slaves']:
+        if slave['hostname'] in hosts:
+            log.info("Reserving all resources on %s" % slave['hostname'])
+            slave_id = slave['id']
+
+            disk = slave['resources']['disk'] - slave['used_resources']['disk']
+            for role in slave['reserved_resources']:
+                disk -= slave['reserved_resources'][role]['disk']
+            reserve(slave_id=slave_id, resource='disk', amount=disk)
```

Comment:
It is because we do a `raise_for_status` in `base_api()` that this becomes a bit more annoying. Since any exception will be caught in this function, which is used by all of the API calls in `mesos_maintenance.py`, we can't easily put error handling specific to dynamic reservations in here.

I think the best approach would be to remove `resp.raise_for_status()` from `base_api()` and move it up two layers into all functions calling `master_api()` (`down`, `up`, `reserve_api`, `unreserve_api`, `maintenance_api`). This will make it possible to adjust how we deal with exceptions for each of the API calls.